SmartDeviceLink/SDLCarWindow.m

```diff
@@ -0,0 +1,213 @@
+//  Copyright © 2016 Ford Motor Company. All rights reserved.
+//
+//  Updated by Joel Fischer, Livio Inc., on 11/27/17.
+
+#import <CommonCrypto/CommonDigest.h>
+#import <ImageIO/ImageIO.h>
+#import <MobileCoreServices/MobileCoreServices.h>
+
+#import "SDLCarWindow.h"
+#import "SDLGlobals.h"
+#import "SDLImageResolution.h"
+#import "SDLLogMacros.h"
+#import "SDLNotificationConstants.h"
+#import "SDLStreamingMediaConfiguration.h"
+#import "SDLStreamingMediaLifecycleManager.h"
+#import "SDLStreamingMediaManagerConstants.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SDLCarWindow ()
+
+@property (weak, nonatomic, nullable) SDLStreamingMediaLifecycleManager *streamManager;
+
+@property (assign, nonatomic) BOOL drawsAfterScreenUpdates;
+
+@property (assign, nonatomic, getter=isLockScreenPresenting) BOOL lockScreenPresenting;
+@property (assign, nonatomic, getter=isLockScreenDismissing) BOOL lockScreenBeingDismissed;
+
+@end
+
+@implementation SDLCarWindow
+
+- (instancetype)initWithStreamManager:(SDLStreamingMediaLifecycleManager *)streamManager drawsAfterScreenUpdates:(BOOL)drawsAfterScreenUpdates {
+    self = [super init];
+    if (!self) { return nil; }
+
+    _streamManager = streamManager;
+    _drawsAfterScreenUpdates = drawsAfterScreenUpdates;
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_didReceiveVideoStreamStarted:) name:SDLVideoStreamDidStartNotification object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_didReceiveVideoStreamStopped:) name:SDLVideoStreamDidStopNotification object:nil];
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_willPresentLockScreenViewController:) name:SDLLockScreenManagerWillPresentLockScreenViewController object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_willDismissLockScreenViewController:) name:SDLLockScreenManagerWillDismissLockScreenViewController object:nil];
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_didPresentLockScreenViewController:) name:SDLLockScreenManagerDidPresentLockScreenViewController object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_didDismissLockScreenViewController:) name:SDLLockScreenManagerDidDismissLockScreenViewController object:nil];
+
+    return self;
+}
+
+- (void)syncFrame {
+    if (!self.streamManager.isVideoConnected || self.streamManager.isVideoStreamingPaused) {
+        return;
+    }
+
+    if (self.isLockScreenPresenting || self.isLockScreenDismissing) {
+        SDLLogD(@"Paused CarWindow, lock screen moving");
+        return;
+    }
+
+    CGRect bounds = self.rootViewController.view.bounds;
+
+    UIGraphicsBeginImageContextWithOptions(bounds.size, YES, 1.0f);
+    [self.rootViewController.view drawViewHierarchyInRect:bounds afterScreenUpdates:self.drawsAfterScreenUpdates];
```

Comment:
I think the captured video could be much sharper. Text is hard to read because of a little halo around each letters' edge.

Reply:
The problem is that bandwidth is very restricted, which means we have to make a tradeoff between framerate and sharpness, and we have to err to framerate right now. It's about as good as it currently gets.

Reply:
This is no longer an issue because the hard data limit was removed.